### DISCUSSION TITLE
What is Information Model vs Semantic-data

### DISCUSSION THREAD

**Discussion by mpostol** [mpostol/OPC-UA-OOI#587] [Ideas]
The term is used in many contexts. One of them is OPC Unified Architecture Information Model. Formally it is defined in [OPC 10000-5: OPC Unified Architecture](https://reference.opcfoundation.org/Core/docs/Part5/), but partially we can find some useful definitions in the [OPC 10000-3: OPC Unified Architecture](https://reference.opcfoundation.org/Core/docs/Part3/).

To make the Information Model interchangeable and meaningful it must be defined using a domain-specific language (DSL).

**Comment by mpostol** [comment]
Before answering the question of what it is first we must concern about the purpose of the Information Model (IM) in machine to machine communication (M2M). M2M is a data exchange scenario where human intervention is impractical or impossible ([2020 M.Postol](https://www.researchgate.net/publication/342286885_Object-Oriented_Internet_Reactive_Interoperability)), e.g. robots cooperation.

My point is that the IM in this scenario is a formal description used to 

- make data validation process possible
- assign a meaning to the data exchanged over the wire

**Comment by mpostol** [comment]
## What is semantic data?

To assign meaning to the data we need a semantics, i.e. a set of rules we can apply to valid data to get information or knowledge we need to monitor or control a process the data comes from.

**Comment by mpostol** [comment]
My point is that the semantics of the data is layered as follows

1. Value encoding \(serialization\), e.g. double, Int32, Boolean,
2. Value meaning, e.g. pressure, temperature, flow, etc.
3. Values relation, e.g. boiler object type containing relation between drum, input pipe, and flow neater.
4. Process – data source, e.g. set of boilers: `Boiler #1`, `Boiler #i`, ...., `Boiler #N` located at…..

Concluding Information Model is all about p. 4. It could be used to:

- validate streams of bits exchanged over the wire
- understand the streams of bits exchanged over the wire

**Comment by mpostol** [comment]
Recently from one of the posts, I learned that OPC UA Information Model design is called data modeling.

**Comment by mpostol** [comment]
## MQTT Sparkplug

This recording from last year provides a good introduction to Sparkplug: <https://www.youtube.com/watch?v=PYYlEktoO-c>

This series published by [HiveMQ](https://www.linkedin.com/company/hivemq-gmbh/), a strategic member of the Sparkplug working group, is a very good introduction to Sparkplug concepts: <https://www.hivemq.com/blog/mqtt-sparkplug-essentials-part-1-introduction/>

**Reply by mpostol** [reply]
Check out [xamarin/mqtt](https://github.com/xamarin/mqtt).

**Reply by mpostol** [reply]
Cannot find the implementation of the `Sparkplug`.

**Comment by fgolra** [comment]
I think OPC UA already offers a DSL.  May be not marketed and presented that way, but the underlying modeling representation & semantics direct towards it.

**Reply by mpostol** [reply]
@fgolra I agree. I can enumerate:

- [UANodeSet - Part 6 Annex F (normative) Information Model XML Schema](https://reference.opcfoundation.org/Core/docs/Part6/F.1/)
- [Part 3 Annex C (normative) Graphical Notation](https://reference.opcfoundation.org/Core/docs/Part3/C.1/)
- [Part 3 Annex B (informative) OPC UA Meta Model in UML](https://reference.opcfoundation.org/Core/docs/Part3/B.1/)
- ModelDesign - custom model maintained in the [OPCFoundation/UA-ModelCompiler](https://github.com/OPCFoundation/UA-ModelCompiler)
- a variety of UML custom dialects coupled with a compiler generating UANodeSet
- a variety of tools with a graphical interface resembling  Part 3 Annex C (normative) Graphical Notation and generating UANodeSet
- a variety of tools (including ASMD) with a graphical interface resembling  Part 3 Annex C (normative) Graphical Notation and using ModelDesign representation

Not sure if it is a complete list. LEt's ask people to review the list.

My concern is that only ModelDesign fulfills partially the requirements of a DSL to be exchanged and reused by independent designers to prepare and expand a model. Unfortunately, it is a custom language and could change without any notice.

**Reply by eoursel** [reply]
Hi @mpostol  
Some ideas. 

-  I would say that the Meta Model (Part 3 annex B)  is not a DSL. 
-  The great question is could we consider the UANodeSet XML Schema and ModelDesign XML Schema as a DSL? Yes in theory, but the problem is that UANodeSet Schema is difficult to use for a data architect without using a modeling tool that is able to produce a semantically valid NodeSet file. There are several technical details that must be described in this XML file which are related to the OPC UA metamodel and are completely irrelevant for a data architect who is now aware of the OPC UA metamodels constraints. If you are doing a workshop with a domain specialist it will be able to describe what is his machine, asset, plant, assembly line but he does not care about some details about the kind of OPC UA relationship between 2 entities (HasComponent, HasProperty, OptionalPlaceHolder ...) and others subtle pure OPC UA technical details. The UA NodeSet schema is more a model exchange format than a DSL. 
- The UML dialects do have the same issue as the UANodeSet format. There are too much detailed to be used by Data Architect without any OPC UA Knowledge. 
- I believe that among the current existing solutions the Microsoft Azure Digital Twin Language is certainly the most relevant solution for a DSL able to support OPC UA models thanks to the work from Erich Barnsted from Microsoft
- My dream is a DSL language based on a W3C Ontology based on RDF dialects with a strong focus on the description of semantics.  I am afraid it is not existing yet. But i think a great subject for a thesis.   

Regards

**Reply by mpostol** [reply]
Hi @eoursel. Thanks for your valuable contribution. 

I agree `Meta Model (Part 3 annex B) is not a DSL.`  It is just a UML model and is not to be used for the design purpose.

Rest is more complicated but still, I agree. Let me just add some at hock notes. My point is that first, we must agree on the definition of the language itself. Next, let's add what the `domain specific`  term means.

I did propose to group languages as follows [Languages grouping](https://d.strumenta.community/t/languages-grouping/1243). Based on this discussion let me assume that we are talking about languages to represent a piece of information (knowledge) in a form that can be read by a computer.  In other words, a computer is a consumer of the outcome. On the other hand, the language must be ready to be used as a design means by a human, i.e. readable and reusable by a designer. 

Under this assumption let me propose the following definition: 

> The language is *three sets*:
>
> 1. alphabet - set of characters we can use
> 1. syntax - set of rules we are applying to check the correctness of the concatenation of the character (text - characters streams)
> 1. semantics - set of rules we are using to associate the correct text and the text meaning.

Ad 1. Computers always use the binary alphabet (only two characters are allowed). Designers (humans) prefer to use an alphabet derived from a native language. A trade-off between these two environments is encoding - a set of rules we can use to convert text into a compliant binary stream and back in a mutually unambiguous manner. In case a graphical language is considered there must be a compiler because the graphical alphabet is generally useless for the automation of data processing using computers. 

Ad. 2. As a result of having encoding (mutually unambiguous relationship between binary and text representations), there is no need to redefine semantic and syntax used in the computer and human environments respectively. My point is that it is the main reason why JSON, XML, YAML, XAML, etc. as a foundation for custom DSL's are so popular. UML is an example of graphical language but to exchange pictures and track changes a domain-specific language based on XML is in daily use. To get pictures a dedicated graphical user interface (GUI) must be supported by an application. 

Ad 3.In the case of programming languages, the semantic rules are strictly observed and the association between the correct text (clauses) is usually more or less formally defined. The metalanguages (BNF, EBNF, ABNF, custom) could change but we can say that the rules are clearly stated - semantic rules are defined in the context of the syntax. For final validation, we can use a compiler. For the compiler, the source text is just input data to be processed. I like to ask when a text becomes a program? The answer is when the compiler doesn't complain.

A real challenge with this definition we must face up when the semantics (knowledge domain) is defined far before the language to be used for its representation. Sometimes we must deal with the problem that the "knowledge domain" is inconsistent internally. Maybe in this case we should use the term `notation` instead of `language`. Not sure - waiting for your proposals.

My point is that **Domain-specific** means that the language is designed (the rules are authored) to be used by designers to describe a knowledge aggregated inside a boundary, for example, covered by a part of the OPC UA Spec. 

### Conclusion

Yes, UANodeSet and ModelDesig are schemas so they may be used as the syntax rules to validate the text representing a model. Unfortunately, both documents don't provide precise definitions of the syntax rules. For example definition of the BrowseName is too general and must be narrowed using the spec but it is possible only indirectly and the definition is ambiguous.  In this context, UANodeSet and ModelDesig are not DSL's for sure.

A semantics rules for text compliant with the UANodeSet we can find in [Part 6 Annex F (normative) Information Model XML Schema](https://reference.opcfoundation.org/Core/docs/Part6/F.1/). In this document we can find the following sentence:

>The description in this Annex only discusses details of the semantics that cannot be captured in the schema document. Types which are self-describing are not discussed.

I am reading it as `ambiguous definitions by design`.

A semantics rules for text compliant with the ModelDesign you may find in the [Information Models Development
](https://commsvr.gitbook.io/ooi/semantic-data-processing/informationmodelsdevelopment). This document is derived from the spec and tries to relax internal inconsistencies hopefully.

My concern is if your dream is a real one. My point is that to designs new DSL the spec must be rewritten partially - OPC UA 2.0. But not sure if the community is ready for that. 

As far as I know, UANodeSet to  DTDL conversion is only a partial (very, very limited) solution but I am going to investigate this problem in detail. Now it is an Azure-dependent solution. I am considering implementing a next converter as a part of [Address Space Prototyping Tool (asp.exe)](https://github.com/mpostol/OPC-UA-OOI/tree/master/SemanticData/AddressSpaceComplianceTestTool#address-space-prototyping-tool-aspexe).

**Comment by mpostol** [comment]
Consider answer from Kevin Moesker at: <https://www.linkedin.com/feed/update/urn:li:activity:6794301780877479936?commentUrn=urn%3Ali%3Acomment%3A%28activity%3A6794301780877479936%2C6796099441259556865%29>

**Comment by mpostol** [comment]
In [The Industrial Internet of Things Vocabulary Technical Report](https://www.iiconsortium.org/vocab/index.htm), Version 2.2
November 6, 2019, there are definitions relevant in my opinion:

| term | Definition | Source
|:-|-|-|
data| content represented in a digital and formalized manner suitable for communication, storage, interpretation or processing| IIC, inspired by ISO/IEC 2382:2015
information| data that within a certain context has a particular meaning|IIC, inspired by ISO/IEC 2382:2015
semantic interoperability| interoperability such that the meaning of the exchanged information can be understood by the participating systems |
